[WIP] Make Trilinos a TPL

aprokop opened the pull request:
Wanted to see how hard that would be. Not too bad, and right now am able to build `Utils` and `Search`, though some Search tests fail.

Some stuff that will need fixing:
- [x] ETI
  ETI is now completely independent from Tpetra in terms of what we instantiate on
- [x] Proper handling of Teuchos unit tests main cpp files
- [ ] Update docs
  Need to indicate that one needs to add `-D Trilinos_ENABLE_INSTALL_CMAKE_CONFIG_FILES=ON` for Trilinos to be picked up
- [ ] Update testing infrastructure
  Docker, Jenkins, etc.

dalg24 commented on packages/Benchmarks/HybridTransport/test/CMakeLists.txt:
```diff
@@ -9,6 +9,7 @@ TRIBITS_ADD_EXECUTABLE_AND_TEST(
   STANDARD_PASS_OUTPUT
   FAIL_REGULAR_EXPRESSION "data race;leak;runtime error"
   )
+TARGET_LINK_LIBRARIES(DataTransferKit_CartesianMesh ${Trilinos_LIBRARIES})
```
That doesn't look right.  I feel like we should be doing `target_link_libraries(dtk <trilinos components>...)` or in that case use `cmake/Dependencies`.

aprokop commented:
@dalg24 I addressed your comments. However, it seems that running cmake configure in Docker fails with unusual (for me) error. Running it on my workstation succeeds in both standalone and combined builds. Would appreciate any help figuring this out.

aprokop commented on packages/Utils/src/CMakeLists.txt:
```diff
@@ -35,4 +25,10 @@ TRIBITS_ADD_LIBRARY(
+TARGET_INCLUDE_DIRECTORIES(
+  ${DTK_UTILS_LIBNAME}
+  PUBLIC
+  "${CMAKE_CURRENT_SOURCE_DIR};${PACKAGE_BINARY_DIR}"  # the latter to find config
```
I know that this is the line where it fails, but I have no idea how to fix it. I did read through that link, and tried adding the generators to no avail.

aprokop commented on CMakeLists.txt:
```diff
@@ -2,127 +2,217 @@
+
+    # Build DataTransferKit shared/static the same way as Trilinos
+    if(Trilinos_BUILD_SHARED_LIBS)
+        set(DataTransferKit_LIBRARY_TYPE SHARED)
```
Hmm, I think it may be used by tribits to change the static/shared type for a particular package build.

aprokop commented:
Rebased on top of master (which includes flattened interface).

aprokop commented:
I think this PR is almost ready. The one thing I'm struggling with and would appreciate help, is the Trilinos TPL thing :)

Specifically, we have few places that mention individual Trilinos packages:
1. In root `CMakeLists.txt`:
   ```cmake
    FIND_PACKAGE(Trilinos REQUIRED
        COMPONENTS Kokkos
        OPTIONAL_COMPONENTS Teuchos Tpetra Intrepid2)
   ```
2. In `TPLsList.cmake`:
   ```cmake                                                                                                                                                                                          
   TRIBITS_REPOSITORY_DEFINE_TPLS(                                                                                                                                                                
     ...  
     Kokkos        TRIBITS_PKG   SS                                                                                                                                                             
     Teuchos       TRIBITS_PKG   SS                                                                                                                                                             
     Tpetra        TRIBITS_PKG   SS                                                                                                                                                             
     Intrepid2     TRIBITS_PKG   SS                                                                                                                          
   )
   ```
3. In `cmake/Dependencies.cmake`:
   ```cmake
    TRIBITS_PACKAGE_DEFINE_DEPENDENCIES(                                                                                                                                                         
    LIB_REQUIRED_TPLS                                                                                                                                                                          
      MPI                                                                                                                                                                                      
      Kokkos                                                                                                                                                                                   
      Teuchos                                                                                                                                                                                  
      Tpetra                                                                                                                                                                                   
      Intrepid2                                                                                                                                                                                
    LIB_OPTIONAL_TPLS                                                                                                                                                                          
      BoostOrg                                                                                                                                                                                 
      Netcdf                                                                                                                                                                                   
    )           
   ```

I'm really struggling understanding the interaction between all this. Right now, things work if one uses `-D CMAKE_PREFIX_PATH="<path_to_installed_trilinos>` during configure. However, if one simply does `-D Trilinos_DIR="<path_to_installed_trilinos>"`, it finds Trilinos but then fails with
```
  Could not find a package configuration file provided by "Kokkos" with any
  of the following names:

    KokkosConfig.cmake
    kokkos-config.cmake

  Add the installation prefix of "Kokkos" to CMAKE_PREFIX_PATH or set
  "Kokkos_DIR" to a directory containing one of the above files.  If "Kokkos"
  provides a separate development package or SDK, be sure it has been
  installed.
```

I would have hoped that `FIND_PACKAGE(Trilinos ...)` would be sufficient. I see it loads all subpackages cmake configurations, but those don't seem to set things like `Kokkos_DIR`.

aprokop commented:
I pushed a workaround for the issue. Essentially, I implemented a simple `FindTPLTrilinos.cmake` that seems to do what we want.

I think the only remaining tasks here are documentation and updating build infrastructure. For that, we need to decide whether we want to support both in-Trilinos and out-Trilinos build for a while, or just switch to out-Trilinos.

dalg24 commented on cmake/TPLs/FindTPLTrilinos.cmake:
```diff
@@ -0,0 +1,13 @@
+)
+GLOBAL_SET(TPL_Trilinos_INCLUDE_DIRS "${Trilinos_INCLUDE_DIRS};${Trilinos_LIBRARY_DIRS}")
+GLOBAL_SET(TPL_Trilinos_LIBRARIES    "${Trilinos_LIBRARIES};${Trilinos_TPL_LIBRARIES}")
+GLOBAL_SET(TPL_Trilinos_LIBRARY_DIRS "${Trilinos_LIBRARY_DIRS};${Trilinos_TPL_LIBRARY_DIRS}")
```
Add comment that you stole that logic from that TriBITS macro for registering TPLs that are marked as `TRIBITS_PKG`.

aprokop replied:
I didn't. I just added these lines after TriBITS complained about them missing.

aprokop commented:
Hmm, it seems that we *do* need to inherit some of the flags Trilinos was compiled with. Specifically, we somehow need to know the CUDA flags and compilers to be able to build with CUDA backend.

I'm not sure what the best way to proceed. It seems over-restrictive to always use the same flags, but on the other hand it's unclear when and how to inherit the flags.

sslattery commented:
We have been doing this in another project by basically getting all of the flags from Kokkos. This has worked well for us although the library we are building is essentially a Kokkos extension so maybe this is overkill: https://github.com/ECP-copa/Cabana/blob/master/cmake/FindKOKKOS.cmake

aprokop commented:
Rebased on top of `master`.

Right now the main issues is when building with cuda. We need to get `nvcc_wrapper` flags.

With @dalg24, we tried creating Kokkos target in root `CMakeLists` and then linking `dtk_utils` (as the library being used everywhere) to it like this:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a9b7b074..26bfe5c3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -51,6 +51,13 @@ ELSE()
     ##---------------------------------------------------------------------------##
 
     IF(${PACKAGE_NAME}_TRILINOS_TPL)
+        # Get Kokkos target
+        FIND_PACKAGE(PkgConfig)
+        LIST(APPEND CMAKE_PREFIX_PATH ${TPL_Trilinos_DIR} /usr/local/cuda-10.0/lib64)
+        pkg_check_modules(KOKKOS REQUIRED
+            IMPORTED_TARGET kokkos)
+        # target_link_libraries(<lib> PkgConfig::KOKKOS)
+
         # The only place to find Kokkos available devices is in
         # `kokkos_generate_settings.cmake`.  They are stored as
         # KOKKOS_GMAKE_DEVICES with short names (Serial/OpenMP/CUDA) separated
diff --git a/packages/Utils/src/CMakeLists.txt b/packages/Utils/src/CMakeLists.txt
index 034cda61..6aa5e4c5 100644
--- a/packages/Utils/src/CMakeLists.txt
+++ b/packages/Utils/src/CMakeLists.txt
@@ -28,3 +28,4 @@ TARGET_INCLUDE_DIRECTORIES(
   "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR};${PACKAGE_BINARY_DIR}>" # the latter to find config
   "$<INSTALL_INTERFACE:include/>"
 )
+TARGET_LINK_LIBRARIES(${DTK_UTILS_LIBNAME} PkgConfig::KOKKOS)
```
It still does not work, and we found few issues:
- In Kokkos' installation package config `kokkos.pc` the libs line is
   `Libs: -L${libdir} -lkokkos -ldl -lcudart -lcuda`, but `-lkokkos` is not the proper library name, Kokkos has `libkokkoscore`, `libkokkoscontainers`, etc.
- PkgConfig also does not find cuda libraries for some reason, so we had to explicitly append cuda path to `CMAKE_PREFIX_PATH`

masterleinad commented:
For me, this pull request doesn't look all that bad. It seems a little bit weird that we explicitly have to find `Kokkos` and finding `Boost` doesn't really work for some reason. Other than that configuring and building works for me. I haven't tried with `Kokkos-3`, though.

Rombur commented on CMakeLists.txt:
```diff
@@ -2,174 +2,229 @@
+
+    IF(${PACKAGE_NAME}_TRILINOS_TPL)
+        # Get Kokkos target
+        FIND_PACKAGE(PkgConfig)
```
Why are still using PkgConfig? I see the (wrong) flags from PkgConfig

Rombur commented on CMakeLists.txt:
```diff
@@ -2,174 +2,229 @@
+    IF(${PACKAGE_NAME}_TRILINOS_TPL)
+        # Get Kokkos target
+        FIND_PACKAGE(PkgConfig)
+        LIST(APPEND CMAKE_PREFIX_PATH ${TPL_Trilinos_DIR} /usr/local/cuda-10.0/lib64)
```
Why do we have the hardcoded cuda path here?

Rombur commented on CMakeLists.txt:
```diff
@@ -2,174 +2,229 @@
+    ## ETI types
+    ##---------------------------------------------------------------------------##
+
+    IF(${PACKAGE_NAME}_TRILINOS_TPL)
```
Why do we have an IF ELSE here?

aprokop replied:
(I think) 
If we run Trilinos as TPL, this is inside the second pass, where we need to grab device types and ETI configs from Trilinos. Otherwise, if DTK is being built as part of Trilinos, we can simply use the cached Tpetra vars.

dalg24 commented:
Close?

Rombur commented:
Yes, the commits have been merged in master
